src/represents.rs

```diff
@@ -0,0 +1,152 @@
+//! Replaces <!--REPRESENTS tagname--> comments with the HTML which appears in a
+//! paragraph of the form:
+//!   <p>The <code>tagname</code> element <span>represents</span> ...</p>
+
+use std::collections::HashMap;
+use std::io;
+use std::rc::Rc;
+
+use crate::dom_utils::NodeHandleExt;
+use html5ever::local_name;
+use html5ever::tendril::StrTendril;
+use markup5ever_rcdom::{Handle, NodeData};
+
+pub struct Processor {
+    /// Map from tag name (as found in the paragraph) to the <span> which
+    /// contains the text "represents".
+    represents: HashMap<StrTendril, Handle>,
+
+    /// List of <!--REPRESENTS--> comments to be replaced, and what tag name
+    /// they correspond to.
+    placeholders: Vec<(Handle, StrTendril)>,
+}
+
+/// Walks from the text node "represents" and finds the tag name and the
+/// span that marks where the description begins, or returns None if that
+/// cannot be found.
+fn find_tag_name(represents_text: &Handle) -> Option<(StrTendril, Handle)> {
+    let span = represents_text
+        .parent_node()
+        .filter(|p| p.is_html_element(&local_name!("span")))?;
+    let p = span
+        .parent_node()
+        .filter(|p| p.is_html_element(&local_name!("p")))?;
+    let children = p.children.borrow();
+    match &children[..] {
+        [a, b, c, d, ..]
+            if a.node_text().as_deref().map(|x| x.trim()) == Some("The")
+                && b.is_html_element(&local_name!("code"))
+                && c.node_text().as_deref().map(|x| x.trim()) == Some("element")
+                && Rc::ptr_eq(d, &span) =>
+        {
+            Some((b.text_content(), span))
+        }
+        _ => None,
+    }
+}
+
+impl Processor {
+    pub fn new() -> Self {
+        Self {
+            represents: HashMap::new(),
+            placeholders: Vec::new(),
+        }
+    }
+
+    /// Should be called for each node the document. Records when it sees a
+    /// <span>represents</span> and which element it is defining
+    pub fn visit(&mut self, node: &Handle) {
+        match node.data {
+            NodeData::Text { ref contents } if contents.borrow().as_ref() == "represents" => {
+                if let Some((tag, span)) = find_tag_name(node) {
+                    self.represents.insert(tag, span);
+                }
+            }
+            NodeData::Comment { ref contents } if contents.starts_with("REPRESENTS ") => {
+                self.placeholders
+                    .push((node.clone(), contents.subtendril(11, contents.len32() - 11)));
```

Comment:
Btw this line means that `<!-- REPRESENTS body-->` would not work, right? (Because of the extra space before "REPRESENTS"). Is this how the current Perl behaves too? I suppose it doesn't matter because all usages of this in the spec are uniform, but I'm just double checking.

Reply:
The current regex is `<!--REPRESENTS ([^>]+)-->`, which also does not accept a space there. In theory we could try to define a more consistent syntax if we wanted to.